File: cpp/3DLib/Camera.cpp
#include "Camera.h"
#include <glm/gtc/matrix_transform.hpp>
#include <glm/gtc/type_ptr.hpp>

Camera::Camera(glm::vec3 position, glm::vec3 upAxis, glm::vec3 forwardAxis, bool rightHanded) :
	m_CameraPosition{ position },
	m_WorldUpAxis{ upAxis },
	m_RightHanded{ rightHanded }
{
	glm::vec3 cameraRight;
	if (m_RightHanded)
	{
		cameraRight = glm::cross(forwardAxis, m_WorldUpAxis);
	}
	else {
		cameraRight = glm::cross(m_WorldUpAxis, forwardAxis);
	}

	m_CameraReference[0] = glm::vec4{ cameraRight,0 };
	m_CameraReference[1] = glm::vec4{ m_WorldUpAxis, 0 };
	m_CameraReference[2] = glm::vec4{ forwardAxis,0 };
}

void Camera::addHorizontalRotation(float dPhi)
{
	m_Phi += dPhi;
}

void Camera::setPhi(float phi)
{
	m_Phi = phi;
}

void Camera::addVerticalRotation(float dAzimuth)
{
	m_Azimuth += dAzimuth;
}

void Camera::update()
{
	// Create the azimuth rotation
	glm::vec3 cameraRight = m_CameraReference[0];
	glm::mat4 azimuthMatrix = glm::rotate(glm::mat4(1.0f), m_Azimuth, cameraRight);
	glm::mat4 pitchedCamera = azimuthMatrix * m_CameraReference;

	// create the rotation around the up axis
	glm::vec3 cameraUp = m_CameraReference[1];
	m_CameraMatrix = glm::rotate(pitchedCamera, m_Phi, cameraUp);
	m_CameraMatrix[3] = glm::vec4{ m_CameraPosition,1 };
}

bool Camera::isRightHanded() const
{
	return m_RightHanded;
}

const glm::mat4& Camera::getCameraMatrix() const
{
	return m_CameraMatrix;
}

File: dotnet/Camera.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ComputeShaderTutorial
{
    using OpenTK.Mathematics;


    public class Camera
    {
        private Vector3 _cameraPosition;
        private Vector3 _worldUpAxis;
        private bool _rightHanded;

        // Rotation angles
        private float _phi;       // Horizontal rotation angle
        private float _azimuth;   // Vertical rotation angle

        // Reference matrix that stores the camera's (right, up, forward) vectors in its columns
        private Matrix4 _cameraReference;

        // Final camera matrix after applying azimuth/phi rotations + translation
        private Matrix4 _cameraMatrix;

        /// <summary>
        /// Constructs a camera with given position, world-up axis, forward axis,
        /// and whether the coordinate system is right-handed.
        /// </summary>
        public Camera(Vector3 position, Vector3 upAxis, Vector3 forwardAxis, bool rightHanded)
        {
            _cameraPosition = position;
            _worldUpAxis = upAxis;
            _rightHanded = rightHanded;

            // Compute the camera right axis
            Vector3 cameraRight;
            if (_rightHanded)
            {
                // Right-handed cross: forward × up
                cameraRight = Vector3.Cross(forwardAxis, _worldUpAxis);
            }
            else
            {
                // Left-handed cross: up × forward
                cameraRight = Vector3.Cross(_worldUpAxis, forwardAxis);
            }

            // Initialize the reference matrix to identity
            _cameraReference = Matrix4.Identity;

            // Store the axes as columns: [cameraRight, upAxis, forwardAxis, translation=0]
            _cameraReference.Column0 = new Vector4(cameraRight, 0.0f);
            _cameraReference.Column1 = new Vector4(_worldUpAxis, 0.0f);
            _cameraReference.Column2 = new Vector4(forwardAxis, 0.0f);
        }

        public void SetPhi(float phi)
        {
            _phi = phi;
        }

        /// <summary>
        /// Updates the final camera matrix by applying the rotations around
        /// the camera's right axis (for azimuth) and the camera's up axis (for phi).
        /// </summary>
        public void Update()
        {
            // 1. Rotation around the camera's right axis (_cameraReference.Column0)
            Vector3 cameraRight = _cameraReference.Column0.Xyz;
            Matrix4 azimuthMatrix = Matrix4.CreateFromAxisAngle(cameraRight, _azimuth);

            // Apply the azimuth rotation
            Matrix4 pitchedCamera = azimuthMatrix * _cameraReference;

            // 2. Rotation around the up axis (in the reference)
            Vector3 cameraUp = _cameraReference.Column1.Xyz;
            Matrix4 upRotation = Matrix4.CreateFromAxisAngle(cameraUp, _phi);

            // Combine them
            _cameraMatrix = upRotation * pitchedCamera;

            // 3. Put the camera position into the 4th column (translation)
            _cameraMatrix.Column3 = new Vector4(_cameraPosition, 1.0f);
        }

        /// <summary>
        /// Gets the final 4×4 camera matrix (orientation + translation).
        /// </summary>
        public Matrix4 GetCameraMatrix()
        {
            return _cameraMatrix;
        }
    }
}
